### DISCUSSION TITLE
Migrate old repository with GIT LFS

### DISCUSSION THREAD

**Discussion by lirantev** [git-lfs/git-lfs#5091] [Q&A] [answered]
Hi git-lfs community, I hope everyone is doing fine :)
I have a question that might be trivial for people how are more familiar with Git then myself, regarding how-to migrate repository X (let's say TFS hosted) to repository Y (GitHub) whereas repository X has git lfs tracked files on it.
Repo X has been around for 3 years and have 500 branches and 1.5k tags. It also contains some really big files (1GB) and plenty of files that are bigger than 100MB.
We picked git-lfs solution to migrate all those big files to git-lfs artifact hosted on JFroj artifactory (using .lfsconfig file).
The thing that I'm find most unclear to me is, on which state repo x supposed to be when running the `git lfs migrate --everything --include` and before using the git push command?
By repository state I mean, should it be either:

1. mirror clone (git clone --mirror) of repository X --> execute git lfs migrate command --> git push --mirror
2. regular clone (git clone) of repository X --> git lfs migrate command --> git push --mirror
    It took me a while to understand this solution isn't good since git clone only get the default branch and the git push --mirror will cause delete all the remote branches that doesn't exist on local, leaving Repository X only with the one default branch
3. regular clone (git clone) of repository X --> manually checkout each and every remote branch --> git lfs migrate command --> git push --mirror
The reason I'm using git push --mirror in this scenario is due to the fact that all the remote branches are now also local, and the mirror will push all refs that may be affected by the git lfs migrate --everything flag

from days and then more days of testing I assume option 3 is the correct one.
Since there isn't any obvious answer how such a migration should be done, I'll really appreciate any comment.
Also, once the repo is ready to be pushed to GitHub, will just git clone --mirror repository x --> git push --mirror repository Y (GitHub) will be sufficient?
Thanks a lot and sorry for all the words :)
Liran.

**Comment by bk2204** [comment, marked as answer]
Hey,

If repository X already has LFS files, then you should use `git clone --mirror --bare` and then `git lfs fetch --all`.  The latter is not needed if it doesn't already have LFS files.

Once you've done that, you can do `git lfs migrate import` with the `--everything` flag plus whatever `--include` values you need on the local clone.  You can then do a `git lfs push --all` to remote Y and a `git push --no-verify --mirror` to remote Y.  The reason I suggest this order is because normally Git LFS walks the history to find all of the LFS files to push, and since this is very expensive when the remote has no data, it's much easier to simply do the LFS push first and then the Git push, which will avoid that.

You don't need to check out each and every branch to migrate it; the `--everything` option should be sufficient.

**Comment by lirantev** [comment]
Hi,
Thank you for the quick replay.
Repository X doesn't have any LFS files in the initial state before migrating it to GitHub so from your response I'm guessing the right migration path should be:
`git clone --mirror --bare` --> `git lfs migrate import --everything --include` --> `git lfs push --all #repositoryYurl#` --> `git push --no-verify --mirror #repositoryYurl#` ?
will test it out and will update this topic.
Thanks :)

**Reply by bk2204** [reply]
Yup, that looks correct.  Note that `--include` requires an argument; the manual page explains the patterns it takes.

**Reply by lirantev** [reply]
Hi @bk2204 , it took me some time but indeed your answer did the trick, thank you :)
